c/datasetjson.c

```diff
@@ -1658,52 +1658,35 @@ void respondWithDatasetMetadata(HttpResponse *response) {
     respondWithError(response,HTTP_STATUS_BAD_REQUEST,"No dataset name given");
     return;
   }
-
   char *username = response->request->username;
   int dsnLen = strlen(datasetOrMember);
+  char *percentDecoded = cleanURLParamValue(response->slh, datasetOrMember);
+  char *absDsPath_temp = stringConcatenate(response->slh, "//'", percentDecoded);
+  char *absDsPath = stringConcatenate(response->slh, absDsPath_temp, "'");
+
+  if(!isDatasetPathValid(absDsPath)){
+    respondWithError(response,HTTP_STATUS_BAD_REQUEST,"Invalid dataset path");
+    return;
+  }
+
   int lParenIndex = indexOf(datasetOrMember, dsnLen, '(', 0);
-  char pdsDSN[45];
-  char pdsMember[100];
+  DatasetName dsnName;
+  DatasetMemberName memName;
+  extractDatasetAndMemberName(absDsPath, &dsnName, &memName);
   int memberNameLength = 0;
+
   if (lParenIndex > 0){
-    int rParenIndex = lastIndexOf(datasetOrMember, dsnLen, ')');
-    int encodedMemberNameLength = rParenIndex-lParenIndex;
-    if (encodedMemberNameLength == 1) {
-      respondWithError(response,HTTP_STATUS_BAD_REQUEST,"No member name given within parenthesis");
-      return;
-    }
-    else if (encodedMemberNameLength > 100) {
-      respondWithError(response,HTTP_STATUS_BAD_REQUEST,"Member name too long");
-      return;
-    }
-    else if (rParenIndex == dsnLen-1){/*paren must end the member listing*/
-      memcpy(pdsDSN,datasetOrMember,lParenIndex);
-      pdsDSN[lParenIndex] = '\0';
-      if (decodePercentByte(&datasetOrMember[lParenIndex+1],encodedMemberNameLength-1,pdsMember,&memberNameLength)) {
-        respondWithError(response,HTTP_STATUS_BAD_REQUEST,"Malformed member name");
-        return;
-      }
-      int asteriskPos = indexOf(pdsMember,memberNameLength,'*',0);
-      if (asteriskPos == -1 && memberNameLength < 8) {
-        for (memberNameLength; memberNameLength < 8; memberNameLength++) {
-          pdsMember[memberNameLength] = ' ';
-        }
+    for(memberNameLength; memberNameLength < 8; memberNameLength++){
+      if(memName.value[memberNameLength] == ' ' || memName.value[memberNameLength] == '\0'){
+        break;
       }
     }
-    else{
-      respondWithError(response, HTTP_STATUS_BAD_REQUEST,"Malformed dataset name");
-      return;
-      /*error*/
-    }
-  }
-  char *dsn;
-  if (lParenIndex > 0){
-    dsn = pdsDSN;
-    dsnLen = lParenIndex+1;
-  }
-  else{
-    dsn = datasetOrMember;
+  } else {
+    memberNameLength = 0;
+    memset(memName.value, '\0', sizeof(memName.value));
   }
+
+  nullTerminate(dsnName.value, sizeof(dsnName.value) - 1);
```

Comment:
This seems like a bug. `value` is 44 chars long, if all of them are used, you'll set the last char to `\0`.